ovs/proto_trace.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2017 DigitalOcean.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ovs
+
+import (
+	"encoding"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	// ErrInvalidProtoTrace is returned when the output from
+	// ovs-appctl ofproto/trace is in an unexpected format
+	ErrInvalidProtoTrace  = errors.New("invalid ofproto/trace output")
+	datapathActionsRegexp = regexp.MustCompile(`Datapath actions: (.*)`)
+	initialFlowRegexp     = regexp.MustCompile(`Flow: (.*)`)
+	finalFlowRegexp       = regexp.MustCompile(`Final flow: (.*)`)
+)
+
+const (
+	popvlan  = "popvlan"
+	pushvlan = "pushvlan"
+	drop     = "drop"
+)
+
+type DataPathAction interface {
+	encoding.TextUnmarshaler
+}
+
+func NewDataPathAction(action string) DataPathAction {
+	return &dataPathAction{
+		action: action,
+	}
+}
+
+type dataPathAction struct {
+	action string
+}
+
+func (d *dataPathAction) UnmarshalText(b []byte) error {
+	action := string(b)
+	if _, err := strconv.Atoi(action); err == nil {
+		// data path action is to an output port
+		d.action = action
+		return nil
+	}
+
+	if strings.HasPrefix(action, pushvlan) {
+		d.action = pushvlan
```

Comment:
The value of the vlan getting pushed with this action is lost here. I'm wondering if a single dataPathAction struct is expressive enough to show all of the possibilities. I'm wondering if there should be a different struct for each kind of data path action with fields specific to that type. For example, there would be a PushVlanAction struct with the actual vlan (and maybe pcp, etc) as members of that struct. There would also be an OutputPortAction which would have the port number as a member of the struct. Thoughts?

Another complicated action that I've seen is: `set(eth(src=fe:00:00:00:01:01,dst=96:61:6a:c7:d8:ec))` I'm having trouble figuring out the best way to express that with this. I would do it something like this:

type SetEthernetAction struct {
    Src string
    Dst string
}

However, this breaks the UnmarshalText pattern, so we'd have to do it differently.